python_modules/dagster/dagster/_core/definitions/time_window_partitions.py

```diff
@@ -194,6 +196,45 @@ def start_time_for_partition_key(self, partition_key: str) -> datetime:
         # we make the assumption that the parsed partition key is <= the start datetime
         return next(iter(self._iterate_time_windows(partition_key_dt))).start
 
+    def time_windows_for_partition_keys(
+        self, partition_keys: Sequence[str], _valid_time_windows: Optional[List[TimeWindow]] = None
+    ) -> List[TimeWindow]:
+        """
+        Accepts an optional `_valid_time_windows` argument to make creating multipartitions subsets
+        more performant. This avoids refetching the list of valid time windows for each time window
+        partitions subset.
+        """
+        if len(partition_keys) == 0:
+            return []
+
+        partition_keys = list(set(partition_keys))
+        sorted_key_datetimes = sorted(
+            set(
+                [
+                    pendulum.instance(datetime.strptime(partition_key, self.fmt), tz=self.timezone)
+                    for partition_key in partition_keys
+                ]
+            ),
+            key=lambda dt: dt.timestamp(),
+        )
+        curr_key_i = 0
+        partition_key_time_windows = []
+        for time_window in (
```

Comment:
As mentioned on Slack, my concern here is that iterating over all the valid time windows will be O(total number of partitions). Ideally this function would only be O(`len(partition_keys`)).

Here's an idea of how to accomplish that without invoking `cron_string_iterator` on every partition key: aim to invoke `cron_string_iterator` once for every contiguous range of partition keys in `partition_keys`.

Something like:
```python
        cur_windows_iterator = windows_iterator_for_partition_key(sorted_partition_keys[0])
        for partition_key in sorted_partition_keys:
            next_window = next(cur_windows_iterator)
            if next_window.start.strftime(partitions_def.fmt) == partition_key:
                partition_key_time_windows.append(next_window)
            else:
                cur_windows_iterator = windows_iterator_for_partition_key(partition_key)
                partition_key_time_windows.append(next(cur_windows_iterator))
```

Reply:
I tried out this approach.

On a hourly partitions definition with 100K keys, if i randomly select:
- 1/2  of the partitions to be in the subset it takes 60s to build 
- 7/8 of the partitions to be in the subset it takes 17s to build

With the existing implementation:
- 1/2 of the partitions to be in the subset it 4s to build
- 7/8 of the partitions it takes 6s to build

From the flame graph it seems like most of the time is spent creating the croniter object in the approach you mentioned. my hypothesis is that if a time window iterator has to be created more than ~500 times (e.g. if there are more than 500 ranges in the partition keys) then it will be more performant to do a single scan of all the partition keys rather than create a croniter object for each range.

Image attached is the flamegraph for 1/2 partitions materialized in the subset
![1_in_2_partitions](https://user-images.githubusercontent.com/29110579/211451764-b6aa4235-66eb-408b-acce-d31ed43ee456.svg)

Reply:
This is a very cool investigation. It's really amazing how slow croniter is.

> my hypothesis is that if a time window iterator has to be created more than ~500 times (e.g. if there are more than 500 ranges in the partition keys) then it will be more performant to do a single scan of all the partition keys rather than create a croniter object for each range.

I believe that it's very rare to see a real-life partitions subset with this number of ranges. In most cases, there's just a single range, even when there are thousands of partitions. I think it's likely not worth the effort of maintaining code paths that are optimized for more than like 10 ranges.

Reply:
got it--I also agree that in practice people will try to make sure all their partitions are materialized (or at least have large ranges). I figure if users complain that loading the subsets is really slow then we can add some perf improvements after.